### PR TITLE
Enable DKG issue tracking

### PR DESCRIPTION
**Pull request by joshuef**
This is atop #1232 so need that in first

### CODE REVIEW

File: sn_dysfunction/src/lib.rs [hunk 1]
```diff
@@ -163,6 +171,31 @@ impl DysfunctionDetection {
...
+        trace!("Attempting to remove logged dkg session for {:?}", node_id,);
+        let mut has_removed = false;
+
+        if let Some(entry) = self.dkg_issues.get(node_id) {
+            let v = entry.value();
+
+            // only remove the first instance from the vec
+            let prev = v.write().await.pop_front();
+
+            if prev.is_some() {
```
Comment (maqi) [inline sn_dysfunction/src/lib.rs:186]: just `has_removed = prev.is_some();` shall be enough?

File: sn_dysfunction/src/lib.rs [hunk 2]
```diff
@@ -224,6 +257,25 @@ pub(crate) fn get_mean_of(data: &[f32]) -> Option<f32> {
     }
 }
 
+fn std_deviation(data: &[f32]) -> Option<f32> {
```
Comment (maqi) [inline sn_dysfunction/src/lib.rs:260]: better to have a comment explaining what this function aiming for ?

File: sn_dysfunction/src/lib.rs [hunk 3]
```diff
@@ -224,6 +257,25 @@ pub(crate) fn get_mean_of(data: &[f32]) -> Option<f32> {
...
+    match (get_mean_of(data), data.len()) {
+        (Some(data_mean), count) if count > 0 => {
+            let variance = data
+                .iter()
+                .map(|value| {
+                    let diff = data_mean - *value;
```
Comment (maqi) [inline sn_dysfunction/src/lib.rs:266]: just to confirm `data_mean` is in type `f32` right? and the result of `diff` could be `negative`?

File: sn_node/src/bin/sn_node.rs [hunk 4]
```diff
@@ -110,31 +110,29 @@ fn create_runtime_and_node() -> Result<()> {
...
-            Ok(result) => {
+        match run_node(config).await {
+            Ok(_) => {
                 info!("Node has finished running, no runtime errors were reported");
-                Ok(result)
             }
             Err(error) => {
-                warn!("Node instance finished with an error. Restarting node...");
-                Err(error)
+                warn!("Node instance finished with an error: {error:?}");
```
Comment (maqi) [inline sn_node/src/bin/sn_node.rs:123]: so now this `error` no longer trigger an error to be reported back further?

File: sn_node/src/node/api/dispatcher/background_tasks.rs [hunk 5]
```diff
@@ -25,9 +25,9 @@ const PROBE_INTERVAL: Duration = Duration::from_secs(30);
 #[cfg(feature = "back-pressure")]
 const BACKPRESSURE_INTERVAL: Duration = Duration::from_secs(60);
 const SECTION_PROBE_INTERVAL: Duration = Duration::from_secs(300);
-const LINK_CLEANUP_INTERVAL: Duration = Duration::from_secs(120);
+const LINK_CLEANUP_INTERVAL: Duration = Duration::from_secs(10);
```
Comment (maqi) [inline sn_node/src/node/api/dispatcher/background_tasks.rs:28]: just to confirm this dramatic reduce of the interval is ok for the real usage as well?

File: sn_node/src/node/core/comm/mod.rs [hunk 6]
```diff
@@ -374,27 +374,32 @@ impl Comm {
                         match watcher.await_change().await {
                             SendStatus::Sent => {
                                 successes += 1;
+                                debug!("successs");
                                 break; // we now move to checking next recipient send task..
                             }
                             SendStatus::Enqueued => {
+                                debug!("qqqqq");
```
Comment (maqi) [inline sn_node/src/node/core/comm/mod.rs:381]: to be removed? also for other couple of debug logs in this match block

File: sn_node/src/node/core/comm/mod.rs [hunk 7]
```diff
@@ -444,54 +449,33 @@ impl Comm {
...
-            }
-        };
-
-        res
+        debug!("get or create peersesh::: none, so making oneee");
+        let link = Link::new(*peer, self.our_endpoint.clone(), self.msg_listener.clone());
+        let session = PeerSession::new(link);
+        debug!("get or create peersesh::: about to insert");
+        let _ = self.sessions.insert(*peer, session.clone());
+        debug!("get or create peersesh::: insertedddd");
```
Comment (maqi) [inline sn_node/src/node/core/comm/mod.rs:465]: remove or change the spelling?

File: sn_node/src/node/core/comm/mod.rs [hunk 8]
```diff
@@ -444,54 +449,33 @@ impl Comm {
...
+        session
     }
 
     /// Any number of incoming qp2p:Connections can be added.
     /// We will eventually converge to the same one in our comms with the peer.
     async fn add_incoming(&self, peer: &Peer, conn: qp2p::Connection) {
         {
```
Comment (maqi) [inline sn_node/src/node/core/comm/mod.rs:472]: seems this extra layer to life span control can now be removed?

File: sn_node/src/node/core/comm/mod.rs [hunk 9]
```diff
@@ -527,9 +512,11 @@ impl Comm {
             msg_id,
             recipient,
         );
-
+        debug!("oooooooother oeer get");
```
Comment (maqi) [inline sn_node/src/node/core/comm/mod.rs:515]: remove?
same to the couple of following

File: sn_node/src/node/core/comm/peer_session.rs [hunk 10]
```diff
@@ -89,12 +89,14 @@ impl PeerSession {
...
     }
 
+    #[instrument(skip(self, msg_bytes, msg_priority))]
     pub(crate) async fn send(
         &self,
         msg_id: MsgId,
         msg_priority: i32,
         msg_bytes: Bytes,
     ) -> Result<SendWatcher> {
+        debug!("SENNNNDDD");
```
Comment (maqi) [inline sn_node/src/node/core/comm/peer_session.rs:99]: remove

File: sn_node/src/node/core/comm/peer_session.rs [hunk 11]
```diff
@@ -160,7 +165,9 @@ impl PeerSession {
                 }
             }
 
+            debug!("sendloooooping");
```
Comment (maqi) [inline sn_node/src/node/core/comm/peer_session.rs:168]: remove

File: sn_node/src/node/core/messaging/handling/dkg.rs [hunk 12]
```diff
@@ -280,6 +280,11 @@ impl Node {
     }
 
     pub(crate) async fn handle_dkg_failure(&self, failure_set: DkgFailureSigSet) -> Result<Cmd> {
+        // track those failed participants
+        for name in &failure_set.failed_participants {
+            self.log_dkg_issue(*name).await?;
```
Comment (maqi) [inline sn_node/src/node/core/messaging/handling/dkg.rs:285]: wondering whether this `log_dkg_issue` function can take the list directly, to reduce the `async await`?

File: sn_dysfunction/src/detection.rs [hunk 13]
```diff
@@ -133,6 +143,14 @@ impl DysfunctionDetection {
                 };
                 count
             }
+            IssueType::Dkg => {
+                let count = if let Some(entry) = self.dkg_issues.get(node) {
+                    entry.value().read().await.len()
+                } else {
+                    1
```
Comment (maqi) [inline sn_dysfunction/src/detection.rs:150]: if `self.dkg_issues` doesn't have the record for the node, it shall means `no issue for that node`.
so the `count` shall return with 0 ?
why here return with 1 ?

File: sn_node/src/node/core/comm/mod.rs [hunk 14]
```diff
@@ -88,46 +88,46 @@ impl Comm {
...
             let is_connected = session.is_connected().await;
 
             if !is_connected {
-                peers_to_cleanup.push(*peer);
+                if !retain_peers.contains(peer) {
+                    peers_to_cleanup.push(*peer);
+                }
+
+                dysfunction
+                    .track_issue(peer.name(), sn_dysfunction::IssueType::Communication)
```
Comment (Yoga07) [inline sn_node/src/node/core/comm/mod.rs:111]: We cleanup when we lose connection with a node right? Not sure if it's bad enough to be tracked? 🤔

File: sn_node/src/node/core/comm/mod.rs [hunk 15]
```diff
@@ -374,27 +374,32 @@ impl Comm {
                         match watcher.await_change().await {
                             SendStatus::Sent => {
                                 successes += 1;
+                                debug!("successs");
```
Comment (Yoga07) [inline sn_node/src/node/core/comm/mod.rs:377]: Can be cleaned up? (similar logs below too)

File: sn_node/src/node/core/comm/mod.rs [hunk 16]
```diff
@@ -527,9 +512,11 @@ impl Comm {
             msg_id,
             recipient,
         );
-
+        debug!("oooooooother oeer get");
         let peer = self.get_or_create(&recipient).await;
+        debug!("oooooooother oeer send");
```
Comment (Yoga07) [inline sn_node/src/node/core/comm/mod.rs:517]: More cleanup? 😄

File: sn_node/src/node/core/comm/peer_session.rs [hunk 17]
```diff
@@ -105,11 +107,13 @@ impl PeerSession {
...
             msg_id,
             msg_bytes,
-            retries: 0,
+            retries: 3,
             reporter,
         };
 
+        debug!("msg wueueueuueeeeeeeeeeeeeeeeeeee writeeee");
         let _ = self.msg_queue.write().await.push(job, msg_priority);
+        debug!("msg wueueueuueeeeeeeeeeeeeeeeeee write done");
```
Comment (Yoga07) [inline sn_node/src/node/core/comm/peer_session.rs:116]: Clean ups :)

File: sn_dysfunction/src/detection.rs [hunk 18]
```diff
@@ -170,22 +190,50 @@ impl DysfunctionDetection {
...
+
+        // now we store the z-score
+        let mut final_scores = BTreeMap::default();
+
+        for (name, score) in pre_z_scores {
+            let zscore = match (mean, std_dev) {
+                (Some(mean), Some(std_deviation)) => {
+                    let diff = score - mean;
+
+                    if diff < 1.0 {
```
Comment (maqi) [inline sn_dysfunction/src/detection.rs:222]: could diff being a negative, i.e. score is less than mean ?

File: sn_dysfunction/src/detection.rs [hunk 19]
```diff
@@ -319,6 +379,82 @@ mod tests {
             })
         }
 
+
+        #[test]
+        /// Test that givena  range of nodes and a few bad nodes, that we can reliably detect those nodes
```
Comment (maqi) [inline sn_dysfunction/src/detection.rs:384]: given a range of
 reliably detect bad nodes ?

File: sn_dysfunction/src/detection.rs [hunk 20]
```diff
@@ -319,6 +379,82 @@ mod tests {
...
+        /// TODO: right now this fails, as it appears our dysfunction barrier is TooDamnHigh
+        /// This _feels_ like a more correct test thanu sing the arbitrary counts we've had. Thoughts?
```
Comment (maqi) [inline sn_dysfunction/src/detection.rs:386]: than using 

if the testing scenario is more realistic, than the arbitary counts shall be replaced?

File: sn_dysfunction/src/lib.rs (hunk 3 above)
Comment (joshuef) [reply sn_dysfunction/src/lib.rs:266]: Yeh, std dev goes both ways

File: sn_node/src/bin/sn_node.rs (hunk 4 above)
Comment (joshuef) [reply sn_node/src/bin/sn_node.rs:123]: Yup, we don't throw out from this func here for node fails. We log the error and the async block completes so we go round again and restart the node

File: sn_node/src/node/api/dispatcher/background_tasks.rs (hunk 5 above)
Comment (joshuef) [reply sn_node/src/node/api/dispatcher/background_tasks.rs:28]: Yeh, Ci has been fine with it. It doesn't seem to be causing any issue, and should keep nodes leaner...

File: sn_node/src/bin/sn_node.rs [hunk 21]
```diff
@@ -110,31 +110,29 @@ fn create_runtime_and_node() -> Result<()> {
 
     // start a new runtime for a node.
     let rt = tokio::runtime::Runtime::new()?;
-    let res = rt.block_on(async move {
+    let _res = rt.block_on(async move {
```
[the remark this answers is not in the document]
Comment (joshuef) [reply sn_node/src/bin/sn_node.rs:113]: The result is bogus there. Always Ok cos we just log what actually comes out of the node. And then retry.

File: sn_node/src/node/core/comm/mod.rs (hunk 14 above)
Comment (joshuef) [reply sn_node/src/node/core/comm/mod.rs:111]: If we're not connected, that's an issue for a node... I think. Won't happen too often. And should happen evenly unless bad, I thnk

File: sn_node/src/node/core/mod.rs [hunk 22]
```diff
@@ -292,12 +294,31 @@ impl Node {
...
+    pub(crate) async fn log_knowledge_issue(&self, name: XorName) -> Result<()> {
+        trace!("Logging Knowledge issue in dysfunction");
+        self.dysfunction_tracking
+            .track_issue(name, IssueType::Knowledge)
+            .await
+            .map_err(Error::from)
+    }
+
+    /// Log a dkg issue (ie, an initialised but unfinished dkg round for a given participant)
+    pub(crate) async fn log_dkg_issue(&self, name: XorName) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (joshuef) [reply sn_node/src/node/core/mod.rs:314]: For me they're convenience issues to not have to import `IssueType` all over the node code

File: sn_dysfunction/src/detection.rs (hunk 13 above)
Comment (joshuef) [reply sn_dysfunction/src/detection.rs:150]: it doesn't matter as long as it's the same for all nodes, as we're taking an avg. (and 0 can make multplication trickier later)

File: sn_dysfunction/src/detection.rs (hunk 18 above)
Comment (joshuef) [reply sn_dysfunction/src/detection.rs:222]: yep. But in this case the node is behaving too well, so we don't need to worry

Comment (joshuef) [reply sn_dysfunction/src/detection.rs:222]: ill add a comment

File: sn_dysfunction/src/detection.rs (hunk 20 above)
Comment (joshuef) [reply sn_dysfunction/src/detection.rs:386]: udpated

**Review by Yoga07 — APPROVED** [review]

**Review by maqi — APPROVED** [review]

**Pull request merged**